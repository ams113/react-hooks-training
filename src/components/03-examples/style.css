.title {
    color: darkgreen;
}

.quotes {
    color:grey;
}

.text-right {
    text-align: right;
}