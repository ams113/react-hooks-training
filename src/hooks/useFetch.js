import { useEffect, useRef, useState } from 'react';

export const useFetch = ( url ) => {

    const initialState = {
        data: null,
        loading: true,
        error: null
    };

    const isMounted = useRef(true);

    const [state, setState] = useState( initialState );

    useEffect( () => {
        return () => {
            isMounted.current(false);
        }
    }, []);

    useEffect(() => {

        setState( initialState );
        
        fetch( url )
            .then( resp => resp.json() )
            .then( data => {

                if ( isMounted.current ) {
                
                    setState(
                        {
                            loading: false,
                            error: null,
                            data
                    });                   
                }

                return () => state;
            })
            .catch( () => {
                setState(
                    {
                        data: null,
                        loading: false,
                        error: 'Can not load the info'
                });
            });
        
    }, [url]);


    return state;
  
}
