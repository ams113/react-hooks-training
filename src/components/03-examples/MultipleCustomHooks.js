import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import './style.css';

export const MultipleCustomHooks = () => {

    const { count, increment } = useCounter(1);

    const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${count}`);
    const {author, quote } = !!data && data[0];
    

    return (
        <>
          
            <h1 className="title">
                BreakingBad&nbsp;
                 <span className="quotes">Quotes</span>
            </h1>
            <hr />

            {
                loading ? (

                    <div className="alert alert-success text-center">
                        Loading...
                    </div>
                )
                :  (
                    <>
                        <blockquote className="blockquote text-right"> 
                        <p> {quote} </p>
                        <footer className="blockquote-footer"> {author} </footer>
                        </blockquote>
                            
                        <button 
                            className="btn btn-outline-warning"
                            onClick={ () => increment() }
                        >
                            Next quote
                        </button>
                    </>
                )

            }

                    
         
        </>
    )
}
