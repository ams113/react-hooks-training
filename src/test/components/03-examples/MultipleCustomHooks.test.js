import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from "../../../components/03-examples/multipleCustomHooks";
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';

//? Mocks
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Testing Components: MultipleCustomHooks', () => {

    beforeEach( () => { 

        useCounter.mockReturnValue({
            count: 10,
            increment: () => {}
        });

    });

    
    test('should renders correctly', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });


        const wrapper = shallow( <MultipleCustomHooks /> );

        expect( wrapper ).toMatchSnapshot();
    });

    test('should show info', () => {
        
        useFetch.mockReturnValue({
            data: [{author: 'toni', quote: 'Ya te digo'}],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );
        
        expect( wrapper.find('.alert').exists() ).toBe( false );
        expect( wrapper.find('p').text().trim() ).toBe( 'Ya te digo' );
        expect( wrapper.find('.blockquote-footer').text().trim() ).toBe( 'toni' );

    });
    
    
});
