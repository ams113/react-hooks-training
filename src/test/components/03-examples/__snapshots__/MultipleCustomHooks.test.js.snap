// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Components: MultipleCustomHooks should renders correctly 1`] = `
<Fragment>
  <h1
    className="title"
  >
    BreakingBadÂ 
    <span
      className="quotes"
    >
      Quotes
    </span>
  </h1>
  <hr />
  <div
    className="alert alert-success text-center"
  >
    Loading...
  </div>
</Fragment>
`;
